File: gateway/src/providers/onboard.ts
import type { GatewayCommand, GatewayResponse } from "../contracts/commands";
import type { DaemonClient, DaemonAgentState, RequestContext } from "../daemon/client";
import { DaemonClientError } from "../daemon/client";

/**
 * Session states for the interactive onboard flow.
 */
export type OnboardStep = "idle" | "agent_selected" | "env_configured" | "installing" | "done";

export type OnboardSession = {
  step: OnboardStep;
  selectedAgent?: string;
  envVars: Map<string, string>;
};

/**
 * In-memory store for onboard session state.
 * Keyed by `${provider}:${chatId}`.
 */
export class OnboardStore {
  private sessions = new Map<string, OnboardSession>();

  getSession(key: string): OnboardSession | undefined {
    return this.sessions.get(key);
  }

  startSession(key: string): OnboardSession {
    const session: OnboardSession = { step: "idle", envVars: new Map() };
    this.sessions.set(key, session);
    return session;
  }

  updateSession(key: string, update: Partial<OnboardSession>): OnboardSession {
    const session = this.sessions.get(key);
    if (!session) {
      throw new Error("no active session");
    }
    Object.assign(session, update);
    return session;
  }

  clearSession(key: string): void {
    this.sessions.delete(key);
  }

  hasActiveSession(key: string): boolean {
    const session = this.sessions.get(key);
    return !!session && session.step !== "idle" && session.step !== "done";
  }

  // Legacy compat helpers
  setEnv(key: string, envName: string, envValue: string): void {
    let session = this.sessions.get(key);
    if (!session) {
      session = this.startSession(key);
    }
    session.envVars.set(envName, envValue);
  }

  getEnv(key: string): Map<string, string> {
    return this.sessions.get(key)?.envVars ?? new Map();
  }

  clearEnv(key: string): void {
    const session = this.sessions.get(key);
    if (session) {
      session.envVars = new Map();
    }
  }
}

export type OnboardDependencies = {
  daemon: DaemonClient;
  onboardStore: OnboardStore;
};

/**
 * Handle `/onboard` as an interactive step-by-step flow.
 *
 * - `/onboard` with no args: starts/continues the interactive session
 * - `/onboard status`: standalone status check (works anytime)
 * - `/onboard cancel`: abort the current session
 * - `/onboard <text>`: routed as user reply in active session
 */
export async function handleOnboardCommand(
  cmd: GatewayCommand,
  deps: OnboardDependencies,
): Promise<GatewayResponse> {
  const ctx: RequestContext = {
    actor: `${cmd.provider}:${cmd.chatId}`,
    requestId: cmd.requestId,
  };
  const sessionKey = `${cmd.provider}:${cmd.chatId}`;
  const [firstArg, ...restArgs] = cmd.args;

  try {
    // `/onboard status` always works regardless of session state
    if (firstArg === "status") {
      return await handleStatus(cmd.requestId, deps.daemon, ctx);
    }

    // `/onboard cancel` aborts an active session
    if (firstArg === "cancel") {
      return handleCancel(cmd.requestId, sessionKey, deps.onboardStore);
    }

    // No args → start a new session (show welcome, transition to awaiting agent selection)
    if (!firstArg) {
      return await startInteractiveSession(cmd.requestId, sessionKey, deps);
    }

    // If there's an active session, route the input as a reply
    const session = deps.onboardStore.getSession(sessionKey);
    if (session && session.step !== "idle" && session.step !== "done") {
      return await handleSessionReply(cmd.requestId, sessionKey, cmd.args, deps, ctx);
    }

    // No active session and unrecognized arg → treat as agent selection shortcut
    // Start session and immediately select the agent
    deps.onboardStore.startSession(sessionKey);
    deps.onboardStore.updateSession(sessionKey, { step: "idle" });
    return await handleSessionReply(cmd.requestId, sessionKey, cmd.args, deps, ctx);
  } catch (error) {
    if (error instanceof DaemonClientError) {
      return {
        requestId: cmd.requestId,
        result: "error",
        errorCode: error.code,
        message: error.message,
      };
    }
    return {
      requestId: cmd.requestId,
      result: "error",
      errorCode: "E_COMMAND_FAILED",
      message: error instanceof Error ? error.message : "unknown error",
    };
  }
}

async function startInteractiveSession(
  requestId: string,
  sessionKey: string,
  deps: OnboardDependencies,
): Promise<GatewayResponse> {
  const ctx: RequestContext = { actor: sessionKey, requestId };
  const agents = await deps.daemon.listAgents(ctx);

  // Start a fresh session
  deps.onboardStore.startSession(sessionKey);

  const lines = ["🚀 Welcome to Carrier! Let's set up your agents.", ""];
  lines.push("Available agents:");
  for (let i = 0; i < agents.length; i++) {
    const a = agents[i];
    const stateTag = a.installState === "installed" ? " [installed]" : "";
    lines.push(`${i + 1}. ${a.name} (${a.id})${stateTag}`);
  }
  lines.push("");
  lines.push("Reply with the agent name to install (e.g. `/onboard openclaw`), or `/onboard cancel` to abort.");

  return {
    requestId,
    result: "ok",
    message: lines.join("\n"),
  };
}

async function handleSessionReply(
  requestId: string,
  sessionKey: string,
  args: string[],
  deps: OnboardDependencies,
  ctx: RequestContext,
): Promise<GatewayResponse> {
  const session = deps.onboardStore.getSession(sessionKey);
  if (!session) {
    // No session, start one
    return await startInteractiveSession(requestId, sessionKey, deps);
  }

  const input = args.join(" ").trim();

  switch (session.step) {
    case "idle":
      // User is selecting an agent
      return await handleAgentSelection(requestId, sessionKey, input, deps, ctx);

    case "agent_selected":
      // User is providing env vars
      return handleEnvInput(requestId, sessionKey, input, deps.onboardStore);

    case "env_configured":
      // User is confirming install
      return await handleConfirmation(requestId, sessionKey, input, deps, ctx);

    case "installing":
      return {
        requestId,
        result: "ok",
        message: "⏳ Installation is in progress. Please wait...",
      };

    case "done":
      return {
        requestId,
        result: "ok",
        message: "✅ Onboarding is complete! Run `/onboard` to set up another agent, or `/onboard status` to check.",
      };

    default:
      return {
        requestId,
        result: "error",
        errorCode: "E_USAGE",
        message: "Unexpected state. Run `/onboard` to start over.",
      };
  }
}

async function handleAgentSelection(
  requestId: string,
  sessionKey: string,
  agentId: string,
  deps: OnboardDependencies,
  ctx: RequestContext,
): Promise<GatewayResponse> {
  if (!agentId) {
    return {
      requestId,
      result: "error",
      errorCode: "E_USAGE",
      message: "Please provide an agent name. Run `/onboard` to see available agents.",
    };
  }

  // Verify the agent exists
  const agents = await deps.daemon.listAgents(ctx);
  const agent = agents.find((a) => a.id === agentId);
  if (!agent) {
    return {
      requestId,
      result: "error",
      errorCode: "E_AGENT_NOT_FOUND",
      message: `Agent "${agentId}" not found. Run \`/onboard\` to see available agents.`,
    };
  }

  deps.onboardStore.updateSession(sessionKey, {
    step: "agent_selected",
    selectedAgent: agentId,
  });

  const lines = [
    `Selected agent: **${agent.name}** (${agent.id})`,
    "",
    "Provide any environment variables as KEY=VALUE pairs (one per message).",
    "When done, reply with `/onboard done`.",
    "To skip env vars, reply `/onboard done` now.",
  ];

  return {
    requestId,
    result: "ok",
    message: lines.join("\n"),
  };
}

function handleEnvInput(
  requestId: string,
  sessionKey: string,
  input: string,
  store: OnboardStore,
): GatewayResponse {
  // "done" transitions to confirmation
  if (input.toLowerCase() === "done") {
    store.updateSession(sessionKey, { step: "env_configured" });
    const session = store.getSession(sessionKey)!;
    const envCount = session.envVars.size;
    const envSummary = envCount > 0
      ? `\nEnvironment variables: ${[...session.envVars.keys()].join(", ")}`
      : "\nNo environment variables set.";

    const lines = [
      `Ready to install **${session.selectedAgent}**?${envSummary}`,
      "",
      "Reply `/onboard yes` to proceed or `/onboard no` to go back.",
    ];

    return {
      requestId,
      result: "ok",
      message: lines.join("\n"),
    };
  }

  // Parse KEY=VALUE
  const eqIdx = input.indexOf("=");
  if (eqIdx <= 0) {
    return {
      requestId,
      result: "error",
      errorCode: "E_USAGE",
      message: 'Please provide env vars as KEY=VALUE, or reply `/onboard done` to continue.',
    };
  }

  const envName = input.slice(0, eqIdx).trim();
  const envValue = input.slice(eqIdx + 1).trim();
  if (!envName || !envValue) {
    return {
      requestId,
      result: "error",
      errorCode: "E_USAGE",
      message: 'Please provide env vars as KEY=VALUE, or reply `/onboard done` to continue.',
    };
  }

  store.setEnv(sessionKey, envName, envValue);
  return {
    requestId,
    result: "ok",
    message: `✅ ${envName} set. Add more variables or reply \`/onboard done\` to continue.`,
  };
}

async function handleConfirmation(
  requestId: string,
  sessionKey: string,
  input: string,
  deps: OnboardDependencies,
  ctx: RequestContext,
): Promise<GatewayResponse> {
  const normalized = input.toLowerCase().trim();

  if (normalized === "no" || normalized === "back") {
    deps.onboardStore.updateSession(sessionKey, { step: "agent_selected" });
    return {
      requestId,
      result: "ok",
      message: "Going back. Provide env vars as KEY=VALUE, or reply `/onboard done` to continue.",
    };
  }

  if (normalized !== "yes" && normalized !== "y") {
    return {
      requestId,
      result: "error",
      errorCode: "E_USAGE",
      message: "Reply `/onboard yes` to install or `/onboard no` to go back.",
    };
  }

  const session = deps.onboardStore.getSession(sessionKey)!;
  const agentId = session.selectedAgent!;

  deps.onboardStore.updateSession(sessionKey, { step: "installing" });

  // Install the agent
  await deps.daemon.installAgent(agentId, ctx);

  // Start the agent
  try {
    await deps.daemon.startAgent(agentId, ctx);
  } catch (error) {
    deps.onboardStore.updateSession(sessionKey, { step: "done" });
    if (error instanceof DaemonClientError) {
      return {
        requestId,
        result: "ok",
        message: `${agentId} installed but failed to start: ${error.message}`,
      };
    }
    throw error;
  }

  // Get status
  const statuses = await deps.daemon.getStatus(agentId, ctx);
  const status = statuses[0];
  const health = status ? status.health : "unknown";

  deps.onboardStore.updateSession(sessionKey, { step: "done" });

  return {
    requestId,
    result: "ok",
    message: `🎉 ${agentId} installed and running (${health}). Onboarding complete!`,
  };
}

function handleCancel(
  requestId: string,
  sessionKey: string,
  store: OnboardStore,
): GatewayResponse {
  const session = store.getSession(sessionKey);
  if (!session || session.step === "idle" || session.step === "done") {
    return {
      requestId,
      result: "ok",
      message: "No active onboarding session to cancel.",
    };
  }

  store.clearSession(sessionKey);
  return {
    requestId,
    result: "ok",
    message: "🚫 Onboarding cancelled. Run `/onboard` to start again.",
  };
}

async function handleStatus(
  requestId: string,
  daemon: DaemonClient,
  ctx: RequestContext,
): Promise<GatewayResponse> {
  const statuses = await daemon.getStatus(undefined, ctx);
  if (statuses.length === 0) {
    return {
      requestId,
      result: "ok",
      message: "No agents configured.",
    };
  }

  const lines = ["Agent status:"];
  for (const s of statuses) {
    const emoji = s.health === "healthy" ? "🟢" : s.runtimeState === "stopped" ? "⚪" : "🔴";
    lines.push(`${emoji} ${s.name} (${s.id}): ${s.installState}, ${s.runtimeState}, health=${s.health}`);
  }

  return {
    requestId,
    result: "ok",
    message: lines.join("\n"),
  };
}

File: gateway/onboard.go
package gateway

import (
	"context"
	"fmt"
	"log"
	"strings"
)

// handleOnboard handles the /onboard command.
func handleOnboard(ctx context.Context, cmd *GatewayCommand, daemon *DaemonClient, store *OnboardStore) GatewayResponse {
	if store == nil {
		store = NewOnboardStore()
	}
	actor := fmt.Sprintf("%s:%s", cmd.Provider, cmd.ChatID)
	sessionKey := actor

	var firstArg string
	if len(cmd.Args) > 0 {
		firstArg = cmd.Args[0]
	}

	// status always works
	if firstArg == "status" {
		return onboardStatus(ctx, cmd.RequestID, daemon, actor)
	}

	// cancel aborts active session
	if firstArg == "cancel" {
		return onboardCancel(cmd.RequestID, sessionKey, store)
	}

	// no args → start interactive session
	if firstArg == "" {
		return onboardStart(ctx, cmd.RequestID, sessionKey, daemon, store, actor)
	}

	// If active session, route as reply
	if store.hasActive(sessionKey) {
		return onboardReply(ctx, cmd.RequestID, sessionKey, cmd.Args, daemon, store, actor)
	}

	// No active session with arg → agent selection shortcut
	store.start(sessionKey)
	return onboardReply(ctx, cmd.RequestID, sessionKey, cmd.Args, daemon, store, actor)
}

func onboardStart(ctx context.Context, requestID, sessionKey string, daemon *DaemonClient, store *OnboardStore, actor string) GatewayResponse {
	agents, err := daemon.ListAgents(ctx, actor, requestID)
	if err != nil {
		return daemonErrResp(requestID, err)
	}
	store.start(sessionKey)

	lines := []string{"🚀 Welcome to Carrier! Let's set up your agents.", ""}
	lines = append(lines, "Available agents:")
	for i, a := range agents {
		tag := ""
		if a.InstallState == "installed" {
			tag = " [installed]"
		}
		lines = append(lines, fmt.Sprintf("%d. %s (%s)%s", i+1, a.Name, a.ID, tag))
	}
	lines = append(lines, "")
	lines = append(lines, "Reply with the agent name to install (e.g. `/onboard openclaw`), or `/onboard cancel` to abort.")
	return GatewayResponse{RequestID: requestID, Result: "ok", Message: strings.Join(lines, "\n")}
}

func onboardReply(ctx context.Context, requestID, sessionKey string, args []string, daemon *DaemonClient, store *OnboardStore, actor string) GatewayResponse {
	sess := store.get(sessionKey)
	if sess == nil {
		return onboardStart(ctx, requestID, sessionKey, daemon, store, actor)
	}
	input := strings.TrimSpace(strings.Join(args, " "))

	switch sess.Step {
	case OnboardIdle:
		return onboardSelectAgent(ctx, requestID, sessionKey, input, daemon, store, actor)
	case OnboardChannelSelect:
		return onboardSelectChannel(requestID, sessionKey, input, store)
	case OnboardChannelToken:
		return onboardCaptureChannelToken(requestID, sessionKey, input, store)
	case OnboardAgentSelected:
		return onboardSelectProvider(requestID, sessionKey, input, store)
	case OnboardProviderSelected:
		return onboardHandleAuth(requestID, sessionKey, input, store)
	case OnboardAuthConfigured:
		return onboardEnvInput(requestID, sessionKey, input, store)
	case OnboardEnvConfigured:
		return onboardConfirm(ctx, requestID, sessionKey, input, daemon, store, actor)
	case OnboardInstalling:
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: "⏳ Installation is in progress. Please wait..."}
	case OnboardDone:
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: "✅ Onboarding is complete! Run `/onboard` to set up another agent, or `/onboard status` to check."}
	default:
		return errResp(requestID, "E_USAGE", "Unexpected state. Run `/onboard` to start over.")
	}
}

func onboardSelectAgent(ctx context.Context, requestID, sessionKey, agentID string, daemon *DaemonClient, store *OnboardStore, actor string) GatewayResponse {
	if agentID == "" {
		return errResp(requestID, "E_USAGE", "Please provide an agent name. Run `/onboard` to see available agents.")
	}
	agents, err := daemon.ListAgents(ctx, actor, requestID)
	if err != nil {
		return daemonErrResp(requestID, err)
	}
	var found *AgentState
	for i := range agents {
		if agents[i].ID == agentID {
			found = &agents[i]
			break
		}
	}
	if found == nil {
		return errResp(requestID, "E_AGENT_NOT_FOUND", fmt.Sprintf("Agent %q not found. Run `/onboard` to see available agents.", agentID))
	}
	store.update(sessionKey, func(s *OnboardSession) {
		s.SelectedAgent = agentID
		s.SelectedAgentName = found.Name
	})
	if isManagedAgent(agentID) {
		store.update(sessionKey, func(s *OnboardSession) {
			s.Step = OnboardChannelSelect
		})
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: renderManagedChannelPrompt(agentID)}
	}
	store.update(sessionKey, func(s *OnboardSession) {
		s.Step = OnboardAgentSelected
	})

	// Build provider list
	return buildProviderListResponse(requestID, found)
}

func onboardSelectChannel(requestID, sessionKey, input string, store *OnboardStore) GatewayResponse {
	sess := store.get(sessionKey)
	if sess == nil {
		return errResp(requestID, "E_USAGE", "No active session. Run `/onboard` to start.")
	}
	if !isManagedAgent(sess.SelectedAgent) {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardAgentSelected })
		return errResp(requestID, "E_USAGE", "Channel selection is only required for PicoClaw/OpenClaw/ZeroClaw in this flow.")
	}
	channel, ok := parseManagedChannel(sess.SelectedAgent, strings.TrimSpace(input))
	if !ok {
		return errResp(requestID, "E_USAGE", "Unsupported channel. Reply `/onboard telegram` to continue.")
	}
	store.update(sessionKey, func(s *OnboardSession) {
		s.SelectedChannel = channel.ID
		s.ChannelToken = ""
		s.ChannelSetupPending = true
		s.Step = OnboardAgentSelected
	})
	name := strings.TrimSpace(sess.SelectedAgentName)
	if name == "" {
		name = managedAgentDisplayName(sess.SelectedAgent)
	}
	if name == "" {
		name = sess.SelectedAgent
	}
	resp := buildProviderListResponse(requestID, &AgentState{ID: sess.SelectedAgent, Name: name})
	resp.Message = strings.TrimSpace(fmt.Sprintf(
		"✅ %s channel selected: **%s** (`%s`).\n\n🔒 Chat onboarding skips bot token input to avoid exposing secrets in chat history.\nConfigure channel token in Web UI after install.\n\n%s",
		managedAgentDisplayName(sess.SelectedAgent),
		channel.Name,
		channel.ID,
		resp.Message,
	))
	return resp
}

func onboardCaptureChannelToken(requestID, sessionKey, input string, store *OnboardStore) GatewayResponse {
	sess := store.get(sessionKey)
	if sess == nil {
		return errResp(requestID, "E_USAGE", "No active session. Run `/onboard` to start.")
	}
	if !isManagedAgent(sess.SelectedAgent) {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardAgentSelected })
		return errResp(requestID, "E_USAGE", "Channel token input is only valid for managed agents.")
	}
	store.update(sessionKey, func(s *OnboardSession) {
		s.ChannelToken = ""
		s.ChannelSetupPending = true
		s.Step = OnboardAgentSelected
	})
	name := strings.TrimSpace(sess.SelectedAgentName)
	if name == "" {
		name = sess.SelectedAgent
	}
	agent := &AgentState{ID: sess.SelectedAgent, Name: name}
	resp := buildProviderListResponse(requestID, agent)
	channelID := strings.TrimSpace(sess.SelectedChannel)
	if channelID == "" {
		channelID = "selected channel"
	}
	resp.Message = strings.TrimSpace(fmt.Sprintf(
		"🔒 Bot token entry in chat onboarding is disabled to protect secrets.\nConfigure token for `%s` in Web UI, then continue.\n\n%s",
		channelID,
		resp.Message,
	))
	return resp
}

func onboardEnvInput(requestID, sessionKey, input string, store *OnboardStore) GatewayResponse {
	if strings.ToLower(input) == "done" {
		sess := store.update(sessionKey, func(s *OnboardSession) {
			s.Step = OnboardEnvConfigured
		})
		keys := make([]string, 0, len(sess.EnvVars))
		for k := range sess.EnvVars {
			keys = append(keys, k)
		}
		envSummary := "\nNo environment variables set."
		if len(keys) > 0 {
			envSummary = "\nEnvironment variables: " + strings.Join(keys, ", ")
		}

		channelLine := ""
		if sess.SelectedChannel != "" {
			channelLine = fmt.Sprintf("\nChannel: %s", sess.SelectedChannel)
			if sess.ChannelSetupPending {
				channelLine += " (token setup pending in Web UI)"
			}
		}
		providerLine := ""
		if sess.SelectedProvider != "" {
			providerLine = fmt.Sprintf("\nLLM Provider: %s", sess.SelectedProvider)
		}

		lines := []string{
			fmt.Sprintf("Ready to install **%s**?%s%s%s", sess.SelectedAgent, channelLine, providerLine, envSummary),
			"",
			"Reply `/onboard yes` to proceed or `/onboard no` to go back.",
		}
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: strings.Join(lines, "\n")}
	}
	eqIdx := strings.IndexByte(input, '=')
	if eqIdx <= 0 {
		return errResp(requestID, "E_USAGE", "Please provide env vars as KEY=VALUE, or reply `/onboard done` to continue.")
	}
	envName := strings.TrimSpace(input[:eqIdx])
	envValue := strings.TrimSpace(input[eqIdx+1:])
	if envName == "" || envValue == "" {
		return errResp(requestID, "E_USAGE", "Please provide env vars as KEY=VALUE, or reply `/onboard done` to continue.")
	}
	store.update(sessionKey, func(s *OnboardSession) {
		s.EnvVars[envName] = envValue
	})
	return GatewayResponse{RequestID: requestID, Result: "ok", Message: fmt.Sprintf("✅ %s set. Add more variables or reply `/onboard done` to continue.", envName)}
}

func onboardConfirm(ctx context.Context, requestID, sessionKey, input string, daemon *DaemonClient, store *OnboardStore, actor string) GatewayResponse {
	normalized := strings.ToLower(strings.TrimSpace(input))
	if normalized == "no" || normalized == "back" {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardAuthConfigured })
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: "Going back. Provide env vars as KEY=VALUE, or reply `/onboard done` to continue."}
	}
	if normalized != "yes" && normalized != "y" {
		return errResp(requestID, "E_USAGE", "Reply `/onboard yes` to install or `/onboard no` to go back.")
	}
	sess := store.get(sessionKey)
	if sess == nil {
		return errResp(requestID, "E_USAGE", "No active session. Run `/onboard` to start.")
	}
	agentID := sess.SelectedAgent
	store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardInstalling })
	setupNotes := []string{}
	if isManagedAgent(agentID) {
		result, err := prepareManagedOnboard(agentID, sess, actor)
		if err != nil {
			store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardEnvConfigured })
			log.Printf("[gateway] onboarding: prepare %s artifacts failed detail=%s", agentID, RedactErrorMessage(err.Error()))
			return errResp(requestID, "E_ENV", fmt.Sprintf("failed to prepare %s onboarding artifacts", agentID))
		}
		setupNotes = append(setupNotes, fmt.Sprintf("%s workspace: %s", managedAgentDisplayName(agentID), result.WorkspacePath))
		setupNotes = append(setupNotes, fmt.Sprintf("%s config: %s", managedAgentDisplayName(agentID), result.ConfigPath))
		setupNotes = append(setupNotes, fmt.Sprintf("Carrier record: %s", result.RecordPath))
		if sess.ChannelSetupPending {
			setupNotes = append(setupNotes, "Channel token setup is pending. Configure it in Web UI before using chat commands.")
		}
	}
	if err := applyOnboardEnvVars(sess.EnvVars); err != nil {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardEnvConfigured })
		log.Printf("[gateway] onboarding: apply env vars failed detail=%s", RedactErrorMessage(err.Error()))
		return errResp(requestID, "E_ENV", "failed to apply environment variables")
	}

	if err := daemon.InstallAgent(ctx, agentID, actor, requestID); err != nil {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardDone })
		return daemonErrResp(requestID, err)
	}
	if err := daemon.StartAgent(ctx, agentID, actor, requestID); err != nil {
		store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardDone })
		if de, ok := err.(*DaemonClientError); ok {
			_, _, msg := mapDaemonErrorToExternal(de.Code)
			return GatewayResponse{RequestID: requestID, Result: "ok", Message: fmt.Sprintf("%s installed but failed to start: %s", agentID, msg)}
		}
		return daemonErrResp(requestID, err)
	}
	statuses, err := daemon.GetStatus(ctx, agentID, actor, requestID)
	health := "unknown"
	if err == nil && len(statuses) > 0 {
		health = statuses[0].Health
	}
	store.update(sessionKey, func(s *OnboardSession) { s.Step = OnboardDone })
	pairHint := ""
	if isPicoclawAgent(agentID) {
		if logs, logErr := daemon.GetLogs(ctx, agentID, 120, actor, requestID); logErr == nil && logs != nil {
			if pairCode := extractPairCode(logs.Lines); pairCode != "" {
				pairHint = fmt.Sprintf("PicoClaw pair code: `%s`. Send `/pair %s` in your PicoClaw Telegram bot.", pairCode, pairCode)
			}
		}
		if pairHint == "" {
			pairHint = "Open your PicoClaw Telegram bot and finish any in-bot pairing/authorization prompts to complete onboarding."
		}
	}
	lines := []string{fmt.Sprintf("🎉 %s installed and running (%s). Onboarding complete!", agentID, health)}
	if len(setupNotes) > 0 {
		lines = append(lines, "")
		lines = append(lines, setupNotes...)
	}
	if pairHint != "" {
		lines = append(lines, "")
		lines = append(lines, pairHint)
	}
	return GatewayResponse{
		RequestID: requestID,
		Result:    "ok",
		Message:   strings.Join(lines, "\n"),
	}
}

func onboardCancel(requestID, sessionKey string, store *OnboardStore) GatewayResponse {
	sess := store.get(sessionKey)
	if sess == nil || sess.Step == OnboardIdle || sess.Step == OnboardDone {
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: "No active onboarding session to cancel."}
	}
	store.clear(sessionKey)
	return GatewayResponse{RequestID: requestID, Result: "ok", Message: "🚫 Onboarding cancelled. Run `/onboard` to start again."}
}

func onboardStatus(ctx context.Context, requestID string, daemon *DaemonClient, actor string) GatewayResponse {
	statuses, err := daemon.GetStatus(ctx, "", actor, requestID)
	if err != nil {
		return daemonErrResp(requestID, err)
	}
	if len(statuses) == 0 {
		return GatewayResponse{RequestID: requestID, Result: "ok", Message: "No agents configured."}
	}
	lines := []string{"Agent status:"}
	for _, s := range statuses {
		emoji := "🔴"
		if s.Health == "healthy" {
			emoji = "🟢"
		} else if s.Runtime == "stopped" {
			emoji = "⚪"
		}
		lines = append(lines, fmt.Sprintf("%s %s (%s): %s, %s, health=%s", emoji, s.Name, s.ID, s.InstallState, s.Runtime, s.Health))
	}
	return GatewayResponse{RequestID: requestID, Result: "ok", Message: strings.Join(lines, "\n")}
}
